### PR TITLE
Implement DFU trigger interface

### PR DESCRIPTION
**Pull request by mathiasap**
- Python-libusb1 used to communicate with trigger interface. (libusb1 binaries for Windows and MacOS added)
- Listing usb devices connected. Needed to associate the serial number of a device with the serial port it us connected to.
- Added an option to perform dfu by specifying the serial number of a device, instead of its serial port.

### CODE REVIEW

File: nordicsemi/lister/lister_backend.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+import abc
```
Comment (chunfantasy) [inline nordicsemi/lister/lister_backend.py]: Just curious what `abc` is.

File: nordicsemi/dfu/dfu_transport_serial.py [hunk 2]
```diff
@@ -302,6 +304,38 @@ def try_to_recover():
...
+
+            if not self.__is_device_in_bootloader_mode(device):
+                trigger = DFUTrigger()
+                trigger.enter_bootloader_mode(device)
+                logger.debug("Serial: DFU bootloader was triggered")
+
+                retry_count = 10
+                wait_time_ms = 500
+                for checks in range(retry_count):
+                    logger.debug("Serial: Waiting {} ms for device to enter bootloader {}/{} time".format(500, checks + 1, retry_count))
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_transport_serial.py]: Syntax: strive to keep lines <= 100 chars

File: nordicsemi/dfu/dfu_transport_serial.py [hunk 3]
```diff
@@ -302,6 +304,38 @@ def try_to_recover():
...
+                        break
+
+                trigger.clean()
+            if not self.__is_device_in_bootloader_mode(device):
+                logger.debug("Serial: Device is either not in bootloader mode, or using an unsupported bootloader.")
+
+    def __is_device_in_bootloader_mode(self, device):
+        if not device:
+            return False
+        return device.vendor_id.lower() == '1915' and device.product_id.lower() == '521f'
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_transport_serial.py]: Could add a comment that explains why we are looking for this vid/pid combination

File: nordicsemi/dfu/dfu_trigger.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+# Importing libusb1 library
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_trigger.py]: Does this import procedure work/pass on win/nix/mac?

File: nordicsemi/dfu/dfu_trigger.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    def __init__(self):
+        self.context = usb1.USBContext()
+
+    def clean(self):
+        self.context.close()
+
+    def select_device(self, listed_device):
+        all_devices = self.context.getDeviceList()
+        filtered_devices = [dev for dev in all_devices\
+        if hex(dev.getVendorID())[2:].lower() == listed_device.vendor_id.lower() and \
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_trigger.py]: Redundant backslash (E502)

File: nordicsemi/dfu/dfu_trigger.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+LIBUSB_REQUEST_TYPE_STANDARD = 0x00 << 5
+LIBUSB_REQUEST_TYPE_CLASS = 0x01 << 5
+LIBUSB_REQUEST_TYPE_VENDOR = 0x02 << 5
+LIBUSB_REQUEST_TYPE_RESERVED = 0x03 << 5
+LIBUSB_RECIPIENT_DEVICE = 0x00
+LIBUSB_RECIPIENT_INTERFACE = 0x01
+LIBUSB_RECIPIENT_ENDPOINT = 0x02
+LIBUSB_RECIPIENT_OTHER = 0x03
+
+ReqTypeInterfaceClass = LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE;
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_trigger.py]: Redundant semicolon at end of line

File: nordicsemi/dfu/dfu_trigger.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        for nordic_device in filtered_devices:
+            try:
+                handle = nordic_device.open()
+                SNO = handle.getSerialNumber()
+                handle.close()
+                if (SNO.lower() == listed_device.serial_number.lower()):
+                    return nordic_device
+
+            except usb1.USBErrorNotFound as err:
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_trigger.py]: Could add comment to explain why this error is swallowed.

File: nordicsemi/dfu/dfu_trigger.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                pass
+            except Exception as err: # LIBUSB_ERROR_NOT_SUPPORTED
```
Comment (bihanssen) [inline nordicsemi/dfu/dfu_trigger.py]: Is it intentional to swallow all exceptions except LIBUSB_ERROR_ACCESS?

File: nordicsemi/lister/device_lister.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        elif sys.platform == 'darwin':
+            self.lister_backend = UnixLister()
+        else:
+            self.lister_backend = None
+
+    def enumerate(self):
+        if self.lister_backend:
+            return self.lister_backend.enumerate()
+        return []
+    def get_device(self, get_all = False, **kwargs):
```
Comment (bihanssen) [inline nordicsemi/lister/device_lister.py]: Missing blank line between functions

File: nordicsemi/lister/enumerated_device.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        self.vendor_id = vendor_id
+        self.product_id = product_id
+        self.serial_number = serial_number
+        self.com_ports = []
+        for port in com_ports:
+            self.add_com_port(port)
+
+    def add_com_port(self, port):
+        if sys.platform == 'darwin':
+            port = port.replace('/dev/cu.', '/dev/tty.')
```
Comment (bihanssen) [inline nordicsemi/lister/enumerated_device.py]: Could add comment about why the port renaming is performed

File: nordicsemi/lister/windows/structures.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
+"""
```
Comment (bihanssen) [inline nordicsemi/lister/windows/structures.py]: Could add the copyright header of gwangyi/pysetupdi to the relevant files

File: nordicsemi/dfu/dfu_trigger.py (hunk 4 above)
Comment (mathiasap) [reply nordicsemi/dfu/dfu_trigger.py]: Tested on Windows 10, Windows 7, MacOS Mojave and Ubuntu 18.04. Windows raises an exception when using 64-bit Python (unsupported). Maybe catch this type of error and handle if this happens?

File: nordicsemi/dfu/dfu_transport_serial.py [hunk 12]
```diff
@@ -302,6 +304,43 @@ def try_to_recover():
...
+                trigger.clean()
+            if not self.__is_device_in_bootloader_mode(device):
+                logger.debug("Serial: Device is either not in bootloader mode, or using an unsupported bootloader.")
+
+    def __is_device_in_bootloader_mode(self, device):
+        if not device:
+            return False
+
+        # Return true if nrf bootloader or Jlink interface detected.
+        return (device.vendor_id.lower() == '1915' and device.product_id.lower() == '521f')  \
```
Comment (blasagna) [inline nordicsemi/dfu/dfu_transport_serial.py]: Nordic's USB library has a feature that the VID and/or PID can be changed. Would this restrict users of nrfutil to using the example secure bootloader in the NRF5 SDK (without forking pc-nrfutil and modifying this logic)?

Comment (bihanssen) [reply nordicsemi/dfu/dfu_transport_serial.py]: @blasagna that is correct, the comparison does only look for the official bootloader that is preloaded on the nRF52840 Dongles.

File: nordicsemi/dfu/dfu_trigger.py [hunk 13]
```diff
@@ -124,5 +134,14 @@ def enter_bootloader_mode(self, listed_device):
...
-        raise NordicSemiException("Device did not exit application mode after dfu was triggered. Serial number: {}, product id 0x{}, vendor id: 0x{}\n\n"\
+        raise NordicSemiException("Device did not exit application mode after dfu was triggered. Serial number: {}, product id 0x{}, vendor id: 0x{}\n\n"
         .format(listed_device.serial_number, listed_device.product_id, listed_device.vendor_id))
+
+if __name__ == "__main__":
+    import sys
+    sys.path.insert(0, os.getcwd())
+    from nordicsemi.lister.device_lister import DeviceLister
+    lister = DeviceLister()
+    dev = lister.get_device(serial_number = "E020C93E0B96")
```
[the remark this answers is not in the document]
Comment (mathiasap) [reply nordicsemi/dfu/dfu_trigger.py]: No, this was not supposed to be added to the commit.

File: nordicsemi/dfu/dfu_transport_serial.py (hunk 12 above)
Comment (mathiasap) [reply nordicsemi/dfu/dfu_transport_serial.py]: @blasagna No, but nrfutil will assume that the device is in application mode and look for a trigger interface. If no trigger interface is found, nrfutil still try to dfu the device, using its original port.

**Pull request merged**